osu.Server.Spectator/Services/LegacyIO.cs

```diff
@@ -0,0 +1,180 @@
+{
+    public class LegacyIO : ILegacyIO
+    {
+        private readonly HttpClient httpClient;
+        private readonly ILogger logger;
+
+        public LegacyIO(HttpClient httpClient, ILoggerFactory loggerFactory)
+        {
+            this.httpClient = httpClient;
+            logger = loggerFactory.CreateLogger("LIO");
+        }
+
+        private async Task<string> runLegacyIO(HttpMethod method, string command, dynamic? postObject = null)
+        {
+            int retryCount = 3;
+
+            retry:
+
+            long time = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+            string url = $"{AppSettings.LegacyIODomain}/_lio/{command}{(command.Contains('?') ? "&" : "?")}timestamp={time}";
+
+            string? serialisedPostObject = postObject switch
+            {
+                null => null,
+                string => postObject,
+                _ => JsonSerializer.Serialize(postObject)
+            };
+
+            logger.LogDebug("Performing LIO request to {method} {url} (params: {params})", method, url, serialisedPostObject);
+
+            try
+            {
+                string signature = hmacEncode(url, Encoding.UTF8.GetBytes(AppSettings.SharedInteropSecret));
+
+                var httpRequestMessage = new HttpRequestMessage
+                {
+                    RequestUri = new Uri(url),
+                    Method = method,
+                    Headers =
+                    {
+                        { "X-LIO-Signature", signature },
+                        { "Accept", "application/json" },
+                    },
+                };
+
+                if (serialisedPostObject != null)
+                {
+                    httpRequestMessage.Content = new ByteArrayContent(Encoding.UTF8.GetBytes(serialisedPostObject));
+                    httpRequestMessage.Content.Headers.ContentType = MediaTypeHeaderValue.Parse("application/json");
+                }
+
+                var response = await httpClient.SendAsync(httpRequestMessage);
+
+                if (response.IsSuccessStatusCode)
+                    return await response.Content.ReadAsStringAsync();
+
+                throw await LegacyIORequestFailedException.Create(url, response);
+            }
+            catch (Exception e)
+            {
+                if (retryCount-- > 0)
+                {
+                    logger.LogError(e, "Legacy IO request to {url} failed, retrying ({retries} remaining)", url, retryCount);
+                    Thread.Sleep(1000);
+                    goto retry;
+                }
+
+                throw;
+            }
+        }
+
+        private static string hmacEncode(string input, byte[] key)
+        {
+            byte[] byteArray = Encoding.ASCII.GetBytes(input);
+
+            using (var hmac = new HMACSHA1(key))
+            {
+                byte[] hashArray = hmac.ComputeHash(byteArray);
+                return hashArray.Aggregate(string.Empty, (s, e) => s + $"{e:x2}", s => s);
+            }
```

Comment:
```suggestion
            byte[] hashArray = HMACSHA1.HashData(key, byteArray);
            return hashArray.Aggregate(string.Empty, (s, e) => s + $"{e:x2}", s => s);
```
A bit cleaner IMO (and potentially faster)